README
wmwave 
Homepage: http://www.schuermann.org/~dockapps

wmwave is dockapp for window maker to display statistical information
about a current wirless ethernet connection.

Installation:
=============

Unzip the the distribution  

	tar xvfz wmwave-0-4.tgz

Change directories

	cd wmwave

Say

	make

and then install the program and manpage by

	make install


Running wmwave:
===============

To run wmwave simply type

	wmwave

Use the command line option '-r <time in milliseconds>' to control the
frequency of sampling of the processes. Use -display and -geometry
options in the usual way.



Troubleshooting:
================

wmwave was tested with Redhat 5.2 on Intel architecture machine
with a /proc filesystem, and a wavelan Sliver WEP IEEE ethernet
card. The kernel needs to support radio networking.  


If wmwave doesn't run,  make sure that

1) the kernel supports radio networking

	if not recompile your kernel with
	
		cd /usr/src/linux
		make xconfig
		
	select radio networking and follow the instructions
	in /usr/src/linux/README.
		
2) the wirless ethernet card and its driver is correctly installed


Please report any bugs you might find to  

	dockapps@schuermann.org



Acknowledgments:
================

I have liberally borrowed code from wmtop by Dan Piponi. Thanks!



debian/changelog
wmwave (0.4-11) unstable; urgency=medium

  [ Guido Günther ]
  * Use 3.0 (quilt) format
  * Move Debian delta to debian/patches
  * Switch to dh
  * Switch to debhelper compat level 12
  * Bump standards version

  [ Reiner Herrmann ]
  * Move libraries to the end
    (Closes: #925859)

 -- Guido Günther <agx@sigxcpu.org>  Fri, 25 Oct 2019 23:59:08 +0200

wmwave (0.4-10) unstable; urgency=medium

  * [38afc39] Switch to debhelper level 8
    (Closes: #817725)
  * [59d5acf] Use homepage field

 -- Guido Günther <agx@sigxcpu.org>  Sun, 13 Mar 2016 00:29:58 +0100

wmwave (0.4-9) unstable; urgency=low

  * fix noise and level calculation for dBm. Thanks to Paul Martin for
    feedback and suggestions (Closes: #406947)

 -- Guido Guenther <agx@sigxcpu.org>  Fri, 15 Jun 2007 16:33:33 +0200

wmwave (0.4-8) unstable; urgency=low

  * use iwlib instead of doing the ioctl ourselfes (Closes: #427594, #406947, #225279)
  * update homepage in debian/copyright and debian/control (Closes: #421405)
  * switch to debhelper 5 

 -- Guido Guenther <agx@sigxcpu.org>  Sat, 09 Jun 2007 14:23:36 +0200

wmwave (0.4-7) unstable; urgency=low

  * add patch from Julien BLACHE <jblache@debian.org> to use the proper ioctl
    instead of parsing /proc/net/wireless, this also fixes some problems when
    the driver reports values in dB or displays erratic values like the level
    being higher than the max level - thanks a lot Julien.
  * bump standards version 
  * update email address

 -- Guido Guenther <agx@sigxcpu.org>  Sat, 13 Jan 2007 22:16:01 +0100

wmwave (0.4-6) unstable; urgency=low

  * Build-Depend on the specific x libraries instead of libx11-dev
    (Closes: 346783)
  * Bump standards version to 3.6.2 (no source changes)
  * quote all menu strings 

 -- Guido Guenther <agx@debian.org>  Mon,  9 Jan 2006 10:10:35 +0100

wmwave (0.4-5) unstable; urgency=low

  * recompile against linux-kernel-headers, fixing segfault on 2.6 kernels
    (Closes: #201713)
  * minor source code cleanups

 -- Guido Guenther <agx@debian.org>  Tue, 23 Dec 2003 00:10:32 +0100

wmwave (0.4-4) unstable; urgency=low

  * manpage cleanups by Ludovic Rousseau, thanks (Closes: #216597)
  * fix typo in copyright file
  * sanitize Makfile a bit and handle DH_BUILD_OPTIONS
  * fix possible overflow in iface[]/wmwave.c
  * bump Standards-Version to 3.6.1 

 -- Guido Guenther <agx@debian.org>  Sun, 26 Oct 2003 19:59:59 +0100

wmwave (0.4-3) unstable; urgency=low

  * package taken from Noah L. Meyerhans, thanks.
  * remove CVS/ directories (Closes: #201220)
  * remove executable bits from wmwave.c
  * fix wrong link display by calling SIOCGIWRANGE ioctl (Closes: #151376)
  * fix typo in description (Closes: #125502)
  * echo 4 > debian/compat, adjust debhelper dependency
  * bump standards version to 3.5.6

 -- Guido Guenther <agx@debian.org>  Sun, 13 Jul 2003 22:32:27 +0200

wmwave (0.4-2.1) unstable; urgency=low

  * NMU
  * debian/control: added xlibs-dev to Build-Depends

 -- Branden Robinson <branden@debian.org>  Fri, 13 Jul 2001 16:45:23 -0500

wmwave (0.4-2) unstable; urgency=low

  * Fixed a bug in the entry in the Debian menu system.  (Closes: Bug#99745)
  * Bumped the standards version up to 3.5.4.0

 -- Noah Meyerhans <noahm@debian.org>  Sat,  2 Jun 2001 22:29:54 -0400

wmwave (0.4-1) unstable; urgency=low

  * Initial Release.

 -- Noah Meyerhans <frodo@morgul.net>  Tue, 07 Feb 2001 16:42:03 -0500



wmwave.c
/*
 * wmwave.c -- WindowMaker wireless statistics view dock app
 * Derived by Carsten Schuermann   carsten@schuermann.org
 * http://www.schuermann.org/~carsten
 * from
 * Dan Piponi dan@tanelorn.demon.co.uk
 * http://www.tanelorn.demon.co.uk
 * who derived it
 * from code originally contained in wmsysmon by Dave Clark (clarkd@skynet.ca)
 * This software is licensed through the GNU General Public License.
 * $Log: wmwave.c,v $
 * Revision 1.7  1999/08/20 13:44:21  carsten
 * version 0.4 complete
 *
 * Revision 1.6  1999/08/19 17:58:52  carsten
 * Almost final version
 *
 * Revision 1.5  1999/08/19 13:54:30  carsten
 * done
 *
 * Revision 1.4  1999/08/19 11:14:50  carsten
 * hookup to /proc/net/wirless complete
 *
 * Revision 1.3  1999/08/19 02:39:07  carsten
 * improved design and hooked it up
 *
 * Revision 1.2  1999/08/16 03:45:34  carsten
 * Added dots
 *
 * Revision 1.1  1999/08/15 15:39:18  carsten
 * Added wmwave project to repository
 *
 */


#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <dirent.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <ctype.h>
#include <math.h>
#include <limits.h>
#include <errno.h>
#include <signal.h>

#include <sys/wait.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/time.h>

#include <iwlib.h>

#include <X11/Xlib.h>
#include <X11/xpm.h>
#include <X11/extensions/shape.h>


#include "wmgeneral.h"

#include "wmwave-master.xpm"

char wmwave_mask_bits[64*64];
int wmwave_mask_width = 64;
int wmwave_mask_height = 64;

#define WMWAVE_VERSION "0.4"

int update_rate=100000;

char *ProgName;

time_t curtime;
time_t prevtime;

int mode = 0;    // default: no card detected

void usage(void);
void printversion(void);
void BlitString(char *name, int x, int y);
void BlitNum(int num, int x, int y);
void wmwave_routine(int, char **);
void DrawBar(float percent, int dx, int dy);
void DrawGreenBar(float percent, int dx, int dy);
int  iw_getinf_stats(const char *ifname, struct iw_range *range, struct iw_statistics *stats);

inline void DrawBar(float percent, int dx, int dy) {
  int tx;
  
  tx = (int)(54.0 * (percent * 0.01));
  copyXPMArea(67, 36, tx, 4, dx, dy);
  copyXPMArea(67, 43, 54-tx, 4, dx+tx, dy); 
}


inline void DrawGreenBar(float percent, int dx, int dy) {
  int tx;
  
  tx = (int)(54.0 * (percent * 0.01));
  copyXPMArea(67, 58, tx, 4, dx, dy);
  copyXPMArea(67, 43, 54-tx, 4, dx+tx, dy); 
}

inline void DrawRedDot() {
  copyXPMArea(80, 65, 6, 6, 52, 5);
}

inline void DrawYellowDot() {
  copyXPMArea(86, 65, 6, 6, 52, 5);
}

inline void DrawGreenDot() {
  copyXPMArea(92, 65, 6, 6, 52, 5);
}

inline void DrawEmptyDot() {
  copyXPMArea(98, 65, 6, 6, 52, 5);
}

float min (float x, float y) {
  if (x < y) {return x;}
  else {return y;}
}

void DisplayWireless(void) {

  struct iw_range range;
  struct iw_statistics stats;

  float link;
  float level;
  float noise;


  FILE *wireless;   // File handle for /proc/net/wireless
  char line[255];
  char iface[6];
  char status [3];
  int nwid = 0;
  int crypt = 0;
  int misc = 0;

  if ((wireless = fopen ("/proc/net/wireless", "r")) != NULL)
    {
      fgets(line,sizeof(line),wireless);
      fgets(line,sizeof(line),wireless);
      if (fgets(line,sizeof(line),wireless) == NULL) {
	mode = 0;
      } else {
	sscanf(line,"%s %s %f %f %f %d %d %d",
	       iface,status,&link,&level,&noise,&nwid,&crypt,&misc);
	iface[strlen(iface)-1] = 0; /* remove ':' */
	if( (!iw_getinf_stats(iface, &range, &stats)) || (!range.max_qual.qual) )
		range.max_qual.qual = 100;
	mode = 1;
      }
      fclose(wireless);


      /* dBm values need a special treatment */
      if ((stats.qual.updated & IW_QUAL_DBM) ||
          (stats.qual.level > range.max_qual.level))
	{
	  /* some cards (like the atheros) simply don't give a floor */
	  float noise_floor = range.max_qual.noise ? (float)(range.max_qual.noise - 0x100) : -128.0;
	  float level_floor = range.max_qual.level ? (float)(range.max_qual.level - 0x100) : -128.0;

	  level = (float)(stats.qual.level - 0x100);
	  noise = (float)(stats.qual.noise - 0x100);

	  level = 100.0 * ( 1.0 - (level / level_floor));
	  noise = 100.0 * ( 1.0 - (noise / noise_floor));
	}
      else
	{
	  float max_level = (float)range.max_qual.level;
	  float max_noise = (float)range.max_qual.noise;

	  level = (float)stats.qual.level;
	  noise = (float)stats.qual.noise;

	  level = 100.0 * level / max_level;
	  noise = 100.0 * noise / max_noise;
	}

      if ((stats.qual.updated & IW_QUAL_LEVEL_INVALID)
	  || level < 0.0 || level > 100.0 )
	level = 0.0;

      if ((stats.qual.updated & IW_QUAL_NOISE_INVALID)
	  || noise < 0.0 || noise > 100.0 )
	noise = 0.0;

      if (!(stats.qual.updated & IW_QUAL_QUAL_INVALID)
	  && (abs(stats.qual.qual) <= abs(range.max_qual.qual)))
	{
	  link = stats.qual.qual;
	  link /= (float) range.max_qual.qual;
	  link *= 100.0;
	}
      else
	link = 0.0;

      /* Print channel information, and signal ratio */
      switch (mode) {
      case 1: BlitString("Quality",4,4);
	if (link<=0.1 * range.max_qual.qual) {
		DrawRedDot ();
	} else if (link<=0.2 * range.max_qual.qual) {
		DrawYellowDot ();
	} else {
		DrawGreenDot();
	}
	BlitString("Link     ", 4,18);	
	DrawBar(min(link, 100.0), 4, 27);
	BlitString("Level    ", 4,32);
	DrawGreenBar(min(level, 100.0), 4, 41);
	BlitString("Noise    ", 4,46);
	DrawGreenBar(min(noise, 100.0), 4, 55);
	break;
      case 0: BlitString("NO CARD",4,4);
	DrawEmptyDot();
	BlitString("         ", 4,18);
	DrawBar(0.0, 4, 27);
	BlitString("         ", 4,32);
	DrawGreenBar(0.0, 4, 41);
	BlitString("         ", 4,46);
	DrawGreenBar(0.0, 4, 55);
	break;
      };
    }
  else {
    printf ("Wirless device /proc/net/wireless not found\n"
            "Enable radio networking and recompile your kernel\n");
    exit (0);
  }
}

/* SIGCHLD handler */
void sig_chld(int signo)
{
  waitpid((pid_t) -1, NULL, WNOHANG);
  signal(SIGCHLD, sig_chld);
}

int main(int argc, char *argv[]) {
  int i;
  
  /* Parse Command Line */
  
  signal(SIGCHLD, sig_chld);
  ProgName = argv[0];
  if (strlen(ProgName) >= 5)
    ProgName += (strlen(ProgName) - 5);
  
  for (i=1; i<argc; i++) {
    char *arg = argv[i];
    
    if (*arg=='-') {
      switch (arg[1]) {
      case 'd' :
	if (strcmp(arg+1, "display")) {
	  usage();
	  exit(1);
	}
	break;
      case 'g' :
	if (strcmp(arg+1, "geometry")) {
	  usage();
	  exit(1);
	}
	break;
      case 'v' :
	printversion();
	exit(0);
	break;
      case 'r':
	if (argc > (i+1)) {
	  update_rate = (atoi(argv[i+1]) * 1000);
	  i++;
	}
	break;
      default:
	usage();
	exit(0);
	break;
      }
    }
  }

  wmwave_routine(argc, argv);
  
  return 0;
}

/*
 * Main loop
 */
void wmwave_routine(int argc, char **argv) {
  XEvent Event;
  struct timeval tv={0,0};
  struct timeval last={0,0};
  
  createXBMfromXPM(wmwave_mask_bits, wmwave_master_xpm, wmwave_mask_width, wmwave_mask_height);
  
  openXwindow(argc, argv, wmwave_master_xpm, wmwave_mask_bits, wmwave_mask_width, wmwave_mask_height);
  
  RedrawWindow();
  
  
  while (1) {
    
    curtime = time(0);
    
    if (1) {
      memcpy(&last, &tv, sizeof(tv));
      
      /*
       * Update display
       */
      DisplayWireless();
      
      RedrawWindow();
    }
    
    /*
     * X Events
     */
    while (XPending(display)) {
      XNextEvent(display, &Event);
      switch (Event.type) {
      case Expose:
	RedrawWindow();
	break;
      case DestroyNotify:
	XCloseDisplay(display);
	exit(0);
	break;
      }
    }
    
    usleep(update_rate);
  }
}

/* get stats & range information */
int iw_getinf_stats(const char *ifname, struct iw_range *range, struct iw_statistics *stats)
{
    int 	skfd;
    int		has_range;

    memset(range, 0, sizeof(struct iw_range));
    memset(stats, 0, sizeof(struct iw_statistics));

    if ((skfd = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
	    return 0;
    }
	
    has_range = (iw_get_range_info(skfd, ifname, range) >= 0);

    if(iw_get_stats(skfd, ifname, stats, range, has_range) < 0) {
            close(skfd);
	    return 0;
    };
    close(skfd);
    return 1;
}

/*
 * Blits a string at given co-ordinates
 */
void BlitString(char *name, int x, int y) {
  int	i;
  int	c;
  int	k;
  
  k = x;
  for (i=0; name[i]; i++)
    {
      
      c = toupper(name[i]); 
      if (c >= 'A' && c <= 'Z')
        {   // its a letter
		 c -= 'A';
	copyXPMArea(c * 6, 74, 6, 8, k, y);
	k += 6;
	} else
	  if (c>='0' && c<='9') {   // its a number or symbol
					 c -= '0';
	  copyXPMArea(c * 6, 64, 6, 8, k, y);
	  k += 6;
	  } else {
	    copyXPMArea(5, 84, 6, 8, k, y);
	    k += 6;
	    
	  }
    }
}

void BlitNum(int num, int x, int y) {
  char buf[1024];
  int newx=x;
  
  sprintf(buf, "%03i", num);
  
  BlitString(buf, newx, y);
}

/*
 * Usage
 */
void usage(void) {
  fprintf(stderr, "\nWmwave - Carsten Schuermann <carsten@schuermann.org>  http://www.schuermann.org/~dockapps\n\n");
  fprintf(stderr, "usage:\n");
  fprintf(stderr, "    -display <display name>\n");
  fprintf(stderr, "    -r                        update rate in milliseconds (default:100)\n");
  fprintf(stderr, "\n");
}

/*
 * printversion
 */
void printversion(void) {
  fprintf(stderr, "wmwave v%s\n", WMWAVE_VERSION);
}
